pkg/kn/commands/revision/revision_describe_test.go

```diff
@@ -31,12 +30,8 @@ import (
 )
 
 func fakeRevision(args []string, response *v1alpha1.Revision) (action client_testing.Action, output string, err error) {
-	buf := new(bytes.Buffer)
-	fakeServing := &fake.FakeServingV1alpha1{&client_testing.Fake{}}
-	cmd := NewKnCommand(KnParams{
-		Output:         buf,
-		ServingFactory: func() (serving.ServingV1alpha1Interface, error) { return fakeServing, nil },
-	})
+	knParams := &common.KnParams{}
+	cmd, fakeServing, buf := utils.CreateTestKnCommand(NewRevisionCommand(knParams), knParams)
```

Comment:
why do we need kn command at all in this unit test. if it's testing describe in isolation, we should just create new describe command.

Reply:
> The PR looks good to me, nice job @maximilien !

thx

> * Why do we need the `kn` directory level below `pkg/` ? I think "commands" can be moved up one level and `kn` removed.

Well that existed from start and agree it's a bit redundant but perhaps indicates the fact that this repo could potentially produce multiple CLIs. However, I'll let others chime in. I don't care one way or the other.

> * Can't we put "root.go", "completion.go" and "version.go" directly into the "commands" package ? This would get rid of the 'etc/' and 'core/' directories/packages.

I tried that. Hated that we would just have three files and not indicate enough info. Also the cyclic dependency is trickier I think but I have to try it again... I had many attempts before settling on this last push. Of course, did not keep all of them :(

With `commons` I kind of liked the idea of a "common" place for things used across all commands, but of course the `commands` package has the same implication.

> * Can we please rename the package for service and revision to `command_revision` and `command_service` (or cmd_revision / cmd_service). This would make it clear from the package name that its about a command and dedicated libs (without references to cobra commands) can be named like 'serving' and 'eventing' without adding to much of confusion.

Well, let's discuss this in the meeting or here. Again don't care myself. I can see an argument for and also one against since the `commands` is already there in the path. So:

`pkg/kn/commands/service` vs `pkg/kn/commands/command_service`

If I understood your suggestion correctly...

> After this bigger restructuring, we can still fine-tune and e.g. refactor the `utils/test_helper.go` which imo shouldn't be there, too.

Yup. Unless we can have more (now or planned) in `utils` I agree.

> * Move as much functionality as possible out of the commands in cobra independent libraries (i.e. into the serving package) so that they can be reused easily. E.g. 'KnServiceCreateCommand' contains too much logic (which even tends to grow e.g. when #156 is added). It should be only a thin layer of a more generic lib imo

Yup. I agree and we should look to setup the correct example for service and revision.

> * Refactor tests to become more unit tests. E.g. commands should be tested only that they propagate the proper info to the library calls (e.g. in serving), the library calls should then use the serving mock library. Only a suggestion, but tests currently are not easy to maintain.

100% agree. I'd use Ginkgo/Gomega and Counterfeiter.... But topic for future conversation. But we should remember this. So I will open an issue on this now (if not already present).

> But as mentioned, I would be very happy if this PR gets merged soonish.

Cool.